### ISSUE TITLE
samplecode/wasmi fails to build against new third_party forks

### ISSUE THREAD

**Issue by brenzi** [apache/teaclave-sgx-sdk#129]
I'm trying to build wasmi samplecode while replacing dependencies by the new mesalock-linux forks.

I get the following error:
```
error[E0277]: `std::rc::Rc<wasmi::TableInstance>` cannot be sent between threads safely
  --> src/lib.rs:60:1
   |
60 | / lazy_static!{
61 | |     static ref SPECDRIVER: SgxMutex<SpecDriver> = SgxMutex::new(SpecDriver::new());
62 | | }
   | |_^ `std::rc::Rc<wasmi::TableInstance>` cannot be sent between threads safely
   |
   = help: within `sgxwasm::SpecDriver`, the trait `core::marker::Send` is not implemented for `std::rc::Rc<wasmi::TableInstance>`
   = note: required because it appears within the type `wasmi::TableRef`
   = note: required because it appears within the type `sgxwasm::SpecModule`
   = note: required because it appears within the type `sgxwasm::SpecDriver`
   = note: required because of the requirements on the impl of `core::marker::Sync` for `std::sync::SgxMutex<sgxwasm::SpecDriver>`
   = note: required by `lazy_static::lazy::Lazy`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

This might have to do with the updated wasmi upstream version for the fork, right?

Here's my `git diff` (had to remove wabt-core dependency too because of #128 

```
diff --git a/samplecode/wasmi/enclave/Cargo.toml b/samplecode/wasmi/enclave/Cargo.toml
index 4b5ce0f..46c464b 100644
--- a/samplecode/wasmi/enclave/Cargo.toml
+++ b/samplecode/wasmi/enclave/Cargo.toml
@@ -11,12 +11,15 @@ crate-type = ["staticlib"]
 default = []
 
 [dependencies]
-wasmi = { path = "../../../third_party/wasmi" }
-serde = { path = "../../../third_party/serde-rs/serde/serde" }
-serde_json = { path = "../../../third_party/serde-rs/json"}
+wasmi = { git = "https://github.com/mesalock-linux/wasmi-sgx" }
+serde = { git = "https://github.com/mesalock-linux/serde-sgx" }
+serde_json = { git = "https://github.com/mesalock-linux/serde-json-sgx"}
+
 sgxwasm = { path = "sgxwasm" }
 lazy_static = { version = "1.1.0", features = ["spin_no_std"] }
 
 [target.'cfg(not(target_env = "sgx"))'.dependencies]
-sgx_types = { path = "../../../sgx_types" }
-sgx_tstd = { path = "../../../sgx_tstd" }
+sgx_tstd = { rev = "v1.0.8", git = "https://github.com/baidu/rust-sgx-sdk"}
+sgx_types = { rev = "v1.0.8", git = "https://github.com/baidu/rust-sgx-sdk"}
+
+
diff --git a/samplecode/wasmi/enclave/sgxwasm/Cargo.toml b/samplecode/wasmi/enclave/sgxwasm/Cargo.toml
index e57c7af..6e899e5 100644
--- a/samplecode/wasmi/enclave/sgxwasm/Cargo.toml
+++ b/samplecode/wasmi/enclave/sgxwasm/Cargo.toml
@@ -4,11 +4,10 @@ version = "0.1.0"
 authors = ["Baidu"]
 
 [dependencies]
-wasmi = { path = "../../../../third_party/wasmi" }
-wabt-core = { path = "../../../../third_party/wabt-rs-core" }
-serde = { path = "../../../../third_party/serde-rs/serde/serde" }
-serde_derive = { path = "../../../../third_party/serde-rs/serde/serde_derive" }
+wasmi = { git = "https://github.com/mesalock-linux/wasmi-sgx" }
+serde = { git = "https://github.com/mesalock-linux/serde-sgx" }
+serde_derive = { git = "https://github.com/mesalock-linux/serde-sgx" }
 
 [target.'cfg(not(target_env = "sgx"))'.dependencies]
-sgx_types = { path = "../../../../sgx_types" }
-sgx_tstd = { path = "../../../../sgx_tstd" }
+sgx_tstd = { rev = "v1.0.8", git = "https://github.com/baidu/rust-sgx-sdk"}
+sgx_types = { rev = "v1.0.8", git = "https://github.com/baidu/rust-sgx-sdk"}
diff --git a/samplecode/wasmi/enclave/sgxwasm/src/lib.rs b/samplecode/wasmi/enclave/sgxwasm/src/lib.rs
index 60959d3..538cc3e 100644
--- a/samplecode/wasmi/enclave/sgxwasm/src/lib.rs
+++ b/samplecode/wasmi/enclave/sgxwasm/src/lib.rs
@@ -37,7 +37,7 @@ extern crate sgx_types;
 #[macro_use]
 extern crate sgx_tstd as std;
 extern crate wasmi;
-extern crate wabt_core;
+//extern crate wabt_core;
 
 use std::{i32, i64, u32, u64, f32};
 use std::prelude::v1::*;
@@ -70,8 +70,8 @@ use wasmi::{ModuleInstance,
             Module,
 };
 
-use wabt_core::script;
-use wabt_core::script::{Value};
+//use wabt_core::script;
+//use wabt_core::script::{Value};
 
 extern crate serde;
 #[macro_use]
@@ -156,7 +156,7 @@ impl SpecModule {
         }
     }
 }
-
+/*
 pub fn spec_to_runtime_value(value: Value) -> RuntimeValue {
     match value {
         Value::I32(v) => RuntimeValue::I32(v),
@@ -165,12 +165,12 @@ pub fn spec_to_runtime_value(value: Value) -> RuntimeValue {
         Value::F64(v) => RuntimeValue::F64(v.into()),
     }
 }
-
+*/
 #[derive(Debug)]
 pub enum Error {
     Load(String),
     Start(Trap),
-    Script(script::Error),
+//    Script(script::Error),
     Interpreter(InterpreterError),
 }
 
@@ -180,12 +180,13 @@ impl From<InterpreterError> for Error {
     }
 }
 
+/*
 impl From<script::Error> for Error {
     fn from(e: script::Error) -> Error {
         Error::Script(e)
     }
 }
-
+*/
 const PRINT_FUNC_INDEX: usize = 0;
 
 impl Externals for SpecModule {
```

**Comment by dingelish** [comment]
seems fixed in 2c4d1a0edb4104cc71d9853f8ae47d677e799106

will work on the hosterror support!

**Comment by dingelish** [comment]
HostError has been supported. So closing this issue.
